packages/core/src/modules/dids/methods/sov/SovDidRegistrar.ts

```diff
@@ -0,0 +1,189 @@
+import type { AgentContext } from '../../../../agent'
+import type { IndyEndpointAttrib } from '../../../ledger'
+import type { DidRegistrar } from '../../domain/DidRegistrar'
+import type { DidCreateOptions, DidCreateResult, DidDeactivateResult, DidUpdateResult } from '../../types'
+import type * as Indy from 'indy-sdk'
+
+import { AgentDependencies } from '../../../../agent/AgentDependencies'
+import { InjectionSymbols } from '../../../../constants'
+import { inject, injectable } from '../../../../plugins'
+import { assertIndyWallet } from '../../../../wallet/util/assertIndyWallet'
+import { IndyLedgerService, IndyPoolService } from '../../../ledger'
+import { DidDocumentRole } from '../../domain/DidDocumentRole'
+import { DidRecord, DidRepository } from '../../repository'
+
+import { addServicesFromEndpointsAttrib, sovDidDocumentFromDid } from './util'
+
+@injectable()
+export class SovDidRegistrar implements DidRegistrar {
+  public readonly supportedMethods = ['sov']
+  private didRepository: DidRepository
+  private indy: typeof Indy
+  private indyLedgerService: IndyLedgerService
+  private indyPoolService: IndyPoolService
+
+  public constructor(
+    didRepository: DidRepository,
+    indyLedgerService: IndyLedgerService,
+    indyPoolService: IndyPoolService,
+    @inject(InjectionSymbols.AgentDependencies) agentDependencies: AgentDependencies
+  ) {
+    this.didRepository = didRepository
+    this.indy = agentDependencies.indy
+    this.indyLedgerService = indyLedgerService
+    this.indyPoolService = indyPoolService
+  }
+
+  public async create(agentContext: AgentContext, options: SovDidCreateOptions): Promise<DidCreateResult> {
+    const { alias, role, submitterDid } = options.options
+    const seed = options.secret?.seed
+
+    if (seed && (typeof seed !== 'string' || seed.length !== 32)) {
+      return {
+        didDocumentMetadata: {},
+        didRegistrationMetadata: {},
+        didState: {
+          state: 'failed',
+          reason: 'Invalid seed provided',
+        },
+      }
+    }
+
+    try {
+      // NOTE: we need to use the createAndStoreMyDid method from indy to create the did
+      // If we just create a key and handle the creating of the did ourselves, indy will throw a
+      // WalletItemNotFound when it needs to sign ledger transactions using this did. This means we need
+      // to rely directly on the indy SDK, as we don't want to expose a createDid method just for.
+      // FIXME: once askar/indy-vdr is supported we need to adjust this to work with both indy-sdk and askar
+      assertIndyWallet(agentContext.wallet)
+      const [unqualifiedIndyDid, verkey] = await this.indy.createAndStoreMyDid(agentContext.wallet.handle, {
+        seed,
+      })
+
+      const qualifiedSovDid = `did:sov:${unqualifiedIndyDid}`
+
+      if (!submitterDid.startsWith('did:sov:')) {
```

Comment:
Can we move this check up? This way we call less funcitons before we check the correct state.